### PR TITLE
feat(crypto-js): allow async load of webassembly

### PR DESCRIPTION
**Pull request by richvdh**
It turns out that Google Chrome refuses to initialise the wasm via the synchronous `WebAssembly.Module` constructor, complaining that it is too big. To be fair, it has a point.

Anyway, that means we need to provide a way to load the wasm asynchronously. So, we introduce an `initAsync()` function which applications can call before they do anything else, to load the wasm in the background.

If the app *doesn't* call `initAsync()`, then we load the wasm synchronously the first time a function that accesses the wasm is called.

### CODE REVIEW

File: bindings/matrix-sdk-crypto-js/scripts/build.sh [hunk 1]
```diff
@@ -21,17 +21,12 @@ RUSTFLAGS='-C opt-level=z' WASM_BINDGEN_WEAKREF=1 wasm-pack build --release --ta
...
-# sed on OSX uses different syntax for sed -i, so let's just avoid it.
-sed -e "/^const path = /d" \
-    -e "s@^const bytes =.*@${loadwasm}@" \
-    -e '/Text..coder.*= require(.util.)/d' \
-    pkg/matrix_sdk_crypto_js.js >pkg/matrix_sdk_crypto_js.js.new
+# In the JavaScript, strip out the existing epilogue, and add our new one
+{
+  sed -e '/^const path = /,$d' pkg/matrix_sdk_crypto_js.js
+  cat scripts/epilogue.js
+} >pkg/matrix_sdk_crypto_js.js.new
```
Comment (Hywan) [inline bindings/matrix-sdk-crypto-js/scripts/build.sh:28]:
```suggestion
} > pkg/matrix_sdk_crypto_js.js.new
```

File: bindings/matrix-sdk-crypto-js/scripts/epilogue.d.ts [hunk 2]
```diff
@@ -0,0 +1,8 @@
+/**
+ * Load the webassembly in the background, if it has not already been loaded.
```
Comment (Hywan) [inline bindings/matrix-sdk-crypto-js/scripts/epilogue.d.ts:2]:
```suggestion
 * Load the WebAssembly module in the background, if it has not already been loaded.
```

File: bindings/matrix-sdk-crypto-js/scripts/epilogue.js [hunk 3]
```diff
@@ -0,0 +1,81 @@
+// Initially, 'wasm' is set to a Proxy object which will synchronously load the webassembly and
```
Comment (Hywan) [inline bindings/matrix-sdk-crypto-js/scripts/epilogue.js:1]:
```suggestion
// Initially, 'wasm' is set to a Proxy object which will synchronously load the WebAssembly module and
```

File: bindings/matrix-sdk-crypto-js/scripts/epilogue.js [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    wasm = instance.exports;
+    wasm.__wbindgen_start();
+    inited = true;
+    return wasm;
+}
+
+let initPromise = null;
+
+/**
+ * Load the webassembly in the background, if it has not already been loaded.
```
Comment (Hywan) [inline bindings/matrix-sdk-crypto-js/scripts/epilogue.js:29]:
```suggestion
 * Load the WebAssembly module in the background, if it has not already been loaded.
```

File: bindings/matrix-sdk-crypto-js/scripts/epilogue.js [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+ * @returns {Promise<void>}
+ */
+module.exports.initAsync = function () {
+    if (inited) {
+        return Promise.resolve();
+    }
+    if (!initPromise) {
+        initPromise = Promise.resolve()
+            .then(() => require("./matrix_sdk_crypto_js_bg.wasm.js"))
+            .then(b64 => WebAssembly.instantiate(unbase64(b64), imports))
```
Comment (Hywan) [inline bindings/matrix-sdk-crypto-js/scripts/epilogue.js:42]: Can we use `WebAssembly.instantiateStreaming` instead?

Also I didn't know `require` was a thing for a browser?

**Review by Hywan — APPROVED** [review]
Thanks for solving this!

File: bindings/matrix-sdk-crypto-js/scripts/epilogue.js (hunk 5 above)
Comment (richvdh) [reply bindings/matrix-sdk-crypto-js/scripts/epilogue.js:42]: > Can we use WebAssembly.instantiateStreaming instead?

It's hard to get hold of the wasm as a stream in an environment-independent way. For node.js we could use the `io` functions, but that doesn't work in a browser. In webpack we could maybe get the url for the wasm via https://webpack.js.org/api/module-variables/#importmetaurl but it's going to be very webpack-specific.

(we could transform the byte buffer into a stream, but I don't believe that offers any benefits?)

> Also I didn't know require was a thing for a browser?

It's not, but webpack and browserify implement it for you.

**Pull request merged**